[TensorPipe] Fix timeout computation

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* **#38928 [TensorPipe] Fix timeout computation**
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging



The original code was
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
which meant we'd never break the loop, as that required `std::chrono::steady_clock::now()` to be *smaller* than `std::chrono::steady_clock::now() + kLargeTimeDuration`.

The fixed code looks like:
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
but by staring at it for a second it becomes clear that the code behaves very differently based on whether `timeoutMap_.empty()`, so I think that for better readability we should reflect that in the code, making that `if` the main one. This then allows us to do a timeout-less wait if there are no messages, which avoids the hacky `kLargeTimeDuration`.

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021/)

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021)

lw committed:
[TensorPipe] Fix timeout computation

The original code was
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
which meant we'd never break the loop, as that required `std::chrono::steady_clock::now()` to be *smaller* than `std::chrono::steady_clock::now() + kLargeTimeDuration`.

The fixed code looks like:
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
but by staring at it for a second it becomes clear that the code behaves very differently based on whether `timeoutMap_.empty()`, so I think that for better readability we should reflect that in the code, making that `if` the main one. This then allows us to do a timeout-less wait if there are no messages, which avoids the hacky `kLargeTimeDuration`.

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Fix timeout computation"




The original code was
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
which meant we'd never break the loop, as that required `std::chrono::steady_clock::now()` to be *smaller* than `std::chrono::steady_clock::now() + kLargeTimeDuration`.

The fixed code looks like:
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
but by staring at it for a second it becomes clear that the code behaves very differently based on whether `timeoutMap_.empty()`, so I think that for better readability we should reflect that in the code, making that `if` the main one. This then allows us to do a timeout-less wait if there are no messages, which avoids the hacky `kLargeTimeDuration`.

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021/)

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021)

[ghstack-poisoned]

mrshenli reviewed (approved):
Great catch! Thanks!

osalpekar commented:
I think you could also just move `steady_clock_time_point earliestTimeout = timeoutMap_.begin()->first;` outside the for loop (as it was in an earlier version of the timeouts PR), and then keep everything else as is. I believe that's how it's done in the rpcRetryThread in rpc_agent.cpp. Though this is a good fix as well!

lw commented:
Yes, that would work too, but TBH I like the fact that since `earliestTimeout` is only needed within the loop its scope is in fact the loop's one, and it doesn't "leak" outside of it.

lw committed:
Update on "[TensorPipe] Fix timeout computation"




The original code was
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
which meant we'd never break the loop, as that required `std::chrono::steady_clock::now()` to be *smaller* than `std::chrono::steady_clock::now() + kLargeTimeDuration`.

The fixed code looks like:
```
steady_clock_time_point earliestTimeout = std::chrono::steady_clock::now() + kLargeTimeDuration;
if (!timeoutMap_.empty()) {
  earliestTimeout = timeoutMap_.begin()->first;
}
if (std::chrono::steady_clock::now() >= earliestTimeout) {
  break;
}
timeoutThreadCV_.wait_until(lock, earliestTimeout);
```
but by staring at it for a second it becomes clear that the code behaves very differently based on whether `timeoutMap_.empty()`, so I think that for better readability we should reflect that in the code, making that `if` the main one. This then allows us to do a timeout-less wait if there are no messages, which avoids the hacky `kLargeTimeDuration`.

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021/)

Differential Revision: [D21703021](https://our.internmc.facebook.com/intern/diff/D21703021)

[ghstack-poisoned]
